############################################################################################################################################################
# This script runs pairwise Wald analysis on the entire RNA-seq dataset using a full model that accounts for treatment, stock, and the interaction between #
# the two. This is the recommended analysis pipeline to use if you want to use the data all together and run pairwise-Wald analysis, although this may be  #
# more useful in comparsion with the lists of DE genes generated by either LRT or separate stock pairwise-Wald analysis.                                   #
#                                                                                                                                                          #
# The script also contains code to build PCAs, heatmaps, and volcano plots based on either the full or stock-subsetted dataset.                            #
############################################################################################################################################################

setwd("/Volumes/Renn_RNAseq_2017/transcriptome")

#Import DESeq2 library in R
library("DESeq2")
library("ggplot2")


#Load gene(/transcript) count matrix and labels
#Other variations of the gene count table with various filters are optional for import as well
countData <- as.matrix(read.csv("gene_count_matrix_filt_refseq_LS.csv", row.names="gene_id"))
colData <- read.csv("PHENO_DATA_filt_LS.csv", row.names=1)

#Note: The PHENO_DATA file contains information on each sample, e.g., sex or population. The exact way to import this depends on the format of the file.
#Check all sample IDs in colData are also in CountData and match their orders
all(rownames(colData) %in% colnames(countData))
#[1] TRUE
countData <- countData[, rownames(colData)]
all(rownames(colData) == colnames(countData))
#[1] TRUE

#Create a DESeqDataSet from count matrix and labels
ddsTC_full <- DESeqDataSetFromMatrix(countData = countData, 
                                     colData = colData, design=~ treat)
#Run the default analysis for DESeq2 and generate results table
ddsTC_full <- DESeq(ddsTC_full, test="LRT", reduced = ~ 1)
res <- results(ddsTC_full)
ddsTC_full <- DESeq(ddsTC_full)

alpha = 0.10
sigtab = res[which(res$padj < alpha), ]
write.table(sigtab, file = "ddsTC_LS_refseq_LRT.txt", sep = "\t")

resultsNames(ddsTC_full)
#MA plot
plotMA(res, main="M-A Plot", ylim=c(-10,10))

#Sort by adjusted p-value and display
(resOrdered <- res[order(res$padj), ])


#########B2 vs B14########
# To extract results, choose the variable of interest and the pair of groups you want to test
# Cooks Cutoff dictates whether or not you want to remove outliers
res <- results( ddsTC_full, contrast = c("treat", "B02", "B14"), cooksCutoff = TRUE )

# Filter for your desired alpha value, and extract relevant taxon names
alpha1 = 0.05
alpha2 = 0.10
sigtab = res[which(res$padj < alpha2), ]
sigtab = cbind(as(sigtab, "data.frame"))
head(sigtab)
dim(sigtab)

# List results
sigtab

write.table(sigtab, file = "ddsTC_LS_refseq_B2vB14_p10.txt", sep = "\t")

#########B2 vs R2########
# To extract results, choose the variable of interest and the pair of groups you want to test
# Cooks Cutoff dictates whether or not you want to remove outliers
res <- results( ddsTC_full, contrast = c("treat", "B02", "R02"), cooksCutoff = TRUE )

# Filter for your desired alpha value, and extract relevant taxon names
sigtab = res[which(res$padj < alpha2), ]
sigtab = cbind(as(sigtab, "data.frame"))
head(sigtab)
dim(sigtab)

# List results
sigtab

write.table(sigtab, file = "ddsTC_LS_refseq_B2vR2_p10.txt", sep = "\t")

#########B2 vs R14########
# To extract results, choose the variable of interest and the pair of groups you want to test
# Cooks Cutoff dictates whether or not you want to remove outliers
res <- results( ddsTC_full, contrast = c("treat", "B02", "R14"), cooksCutoff = TRUE )

# Filter for your desired alpha value, and extract relevant taxon names
sigtab = res[which(res$padj < alpha2), ]
sigtab = cbind(as(sigtab, "data.frame"))
head(sigtab)
dim(sigtab)

# List results
sigtab

write.table(sigtab, file = "ddsTC_LS_refseq_B2vR14_p10.txt", sep = "\t")

#########B14 vs R2########
# To extract results, choose the variable of interest and the pair of groups you want to test
# Cooks Cutoff dictates whether or not you want to remove outliers
res <- results( ddsTC_full, contrast = c("treat", "B14", "R02"), cooksCutoff = TRUE )

# Filter for your desired alpha value, and extract relevant taxon names
sigtab = res[which(res$padj < alpha2), ]
sigtab = cbind(as(sigtab, "data.frame"))
head(sigtab)
dim(sigtab)

# List results
sigtab

write.table(sigtab, file = "ddsTC_LS_refseq_B14vR2_p10.txt", sep = "\t")

#########B14 vs R14########
# To extract results, choose the variable of interest and the pair of groups you want to test
# Cooks Cutoff dictates whether or not you want to remove outliers
res <- results( ddsTC_full, contrast = c("treat", "B14", "R14"), cooksCutoff = TRUE )

# Filter for your desired alpha value, and extract relevant taxon names
sigtab = res[which(res$padj < alpha2), ]
sigtab = cbind(as(sigtab, "data.frame"))
head(sigtab)
dim(sigtab)

# List results
sigtab

write.table(sigtab, file = "ddsTC_LS_refseq_B14vR14_p10.txt", sep = "\t")

#########R2 vs R14########
# To extract results, choose the variable of interest and the pair of groups you want to test
# Cooks Cutoff dictates whether or not you want to remove outliers
res <- results( ddsTC_full, contrast = c("treat", "R02", "R14"), cooksCutoff = TRUE )

# Filter for your desired alpha value, and extract relevant taxon names
sigtab = res[which(res$padj < alpha2), ]
sigtab = cbind(as(sigtab, "data.frame"))
head(sigtab)
dim(sigtab)

# List results
sigtab

write.table(sigtab, file = "ddsTC_LS_refseq_R2vR14_p10.txt", sep = "\t")


########## Plot by gene ###########
#plot by gene name, substitute accordingly for any gene of interest
profile <- plotCounts(ddsTC_full, "MSTRG.37168", 
                      intgroup = c("treat","stock"), returnData = TRUE)
ggplot(profile,
       aes(x = treat, y = count, color = stock, group = stock)) + 
  geom_point() + geom_smooth(se = TRUE, method = "loess") +
  theme(text = element_text(size=20)) + xlab("timepoint") 


######################################
####### Lab Stock only LRT ###########
######################################

ddsTC_LRT <- DESeqDataSetFromMatrix(countData = countData, 
                                    colData = colData, design= ~ treat)

# IMPORTANT: reduced model contains terms that you plan to remove from full model to isolate any remaining terms.
ddsTC_LRT <- DESeq(ddsTC_LRT, test="LRT", reduced = ~ 1)
resTC_LRT <- results(ddsTC_LRT)
resTC_LRT$symbol <- mcols(ddsTC_LRT)$treat

resultsNames(ddsTC_LRT)
write.table(resTC_LRT, file = "resTC_LRT_LS_only_full_results.txt", sep = "\t")

sigtab = resTC_LRT[which(resTC_LRT$pvalue < alpha1), ]
head(order(sigtab$padj))
dim(sigtab)
sigtab

write.table(sigtab, file = "dds_LS_LRT_ucorrected_p05_treat_only.txt", sep = "\t")

########## Plot by gene ###########
#plot by gene name, substitute accordingly for any gene of interest
profile <- plotCounts(ddsTC_LRT, "MSTRG.37168", 
                      intgroup = c("treat","stock"), returnData = TRUE)
ggplot(profile,
       aes(x = treat, y = count, color = stock, group = stock)) + 
  geom_point() + geom_smooth(se = TRUE, method = "loess") +
  theme(text = element_text(size=20)) + xlab("timepoint") 

resTC_LRT["MSTRG.37168",]

#resTC_LRT_hunger <- subset(resTC_LRT, rownames(resTC_LRT)=="MSTRG.10958" | rownames(resTC_LRT)=="MSTRG.11494" | rownames(resTC_LRT)=="MSTRG.12014" | rownames(resTC_LRT)=="MSTRG.12931" | rownames(resTC_LRT)=="MSTRG.12933" | rownames(resTC_LRT)=="MSTRG.13768" | rownames(resTC_LRT)=="MSTRG.15205" | rownames(resTC_LRT)=="MSTRG.15206" | rownames(resTC_LRT)=="MSTRG.16839" | rownames(resTC_LRT)=="MSTRG.16954" | rownames(resTC_LRT)=="MSTRG.17268" | rownames(resTC_LRT)=="MSTRG.17504" | rownames(resTC_LRT)=="MSTRG.17917" | rownames(resTC_LRT)=="MSTRG.18114" | rownames(resTC_LRT)=="MSTRG.18320" | rownames(resTC_LRT)=="MSTRG.20197" | rownames(resTC_LRT)=="MSTRG.22044" | rownames(resTC_LRT)=="MSTRG.22629" | rownames(resTC_LRT)=="MSTRG.23148" | rownames(resTC_LRT)=="MSTRG.23508" | rownames(resTC_LRT)=="MSTRG.25774" | rownames(resTC_LRT)=="MSTRG.27521" | rownames(resTC_LRT)=="MSTRG.28840" | rownames(resTC_LRT)=="MSTRG.29123" | rownames(resTC_LRT)=="MSTRG.30486" | rownames(resTC_LRT)=="MSTRG.30858" | rownames(resTC_LRT)=="MSTRG.32928" | rownames(resTC_LRT)=="MSTRG.33881" | rownames(resTC_LRT)=="MSTRG.35779" | rownames(resTC_LRT)=="MSTRG.3738" | rownames(resTC_LRT)=="MSTRG.38481" | rownames(resTC_LRT)=="MSTRG.40659" | rownames(resTC_LRT)=="MSTRG.4103" | rownames(resTC_LRT)=="MSTRG.41351" | rownames(resTC_LRT)=="MSTRG.41356" | rownames(resTC_LRT)=="MSTRG.43087" | rownames(resTC_LRT)=="MSTRG.44096" | rownames(resTC_LRT)=="MSTRG.44245" | rownames(resTC_LRT)=="MSTRG.44247" | rownames(resTC_LRT)=="MSTRG.48687" | rownames(resTC_LRT)=="MSTRG.51383" | rownames(resTC_LRT)=="MSTRG.53265" | rownames(resTC_LRT)=="MSTRG.56422" | rownames(resTC_LRT)=="MSTRG.59894" | rownames(resTC_LRT)=="MSTRG.60586" | rownames(resTC_LRT)=="MSTRG.62578" | rownames(resTC_LRT)=="MSTRG.63346" | rownames(resTC_LRT)=="MSTRG.64277" | rownames(resTC_LRT)=="MSTRG.69154" | rownames(resTC_LRT)=="MSTRG.70389" | rownames(resTC_LRT)=="MSTRG.71405" | rownames(resTC_LRT)=="MSTRG.72372" | rownames(resTC_LRT)=="MSTRG.8776" | rownames(resTC_LRT)=="gene17037")
#write.table(resTC_LRT_hunger, file = "resTC_LRT_LS_only_hunger.txt", sep = "\t")


#################################################
######## Expression Profile Clustering ##########
#################################################

library( "genefilter" )
library("pheatmap")
# load required packages
if (!require("gplots")) {
  install.packages("gplots", dependencies = TRUE)
  library(gplots)
}
if (!require("RColorBrewer")) {
  install.packages("RColorBrewer", dependencies = TRUE)
  library(RColorBrewer)
}



############
# Plotting #
############


#Regularized log transformation of DESeq2 dataset for better plotting
rld <- rlog( ddsTC_full )

#Variance Stabilizing Transformation (VST) of DESeq2 dataset for better plotting and other downtream applications
vstDESeq <- vst( ddsTC_full ) # you can substitute this dataset for the rld object below, results will be similar

# If you want to export log-transformed data to file run following 2 commands
#rld_mat <- assay(rld)
#write.table(rld_mat, file = "ddsTC_full_log.csv", sep = ",")

# If you want to export VST-transformed data to file run following 2 commands
#vst_Mat <- assay(vstDESeq)
#write.table(vst_mat, file = "ddsTC_full_vst.csv", sep = ",")

#Get N top varying genes not super useful if you're interested in specific gene subset, but good to get general sense of most variable genes
topVarGenes <- head( order( rowVars( assay(rld) ), decreasing=TRUE ), 100 )
matTop <- assay(rld)[ topVarGenes, ]
matTop <- matTop - rowMeans(matTop)

#Get genes significant at padj < 0.10
DEgenes1 <- which(resTC_LRT$padj < alpha2)
matDE1 <- assay(rld)[ DEgenes1, ]
matDE1 <- matDE1 - rowMeans(matDE1)

#Get genes significant at unadjusted-p < 0.05
DEgenes2 <- which(resTC_LRT$pvalue < alpha1)
matDE1 <- assay(rld)[ DEgenes2, ]
matDE2 <- matDE2 - rowMeans(matDE2)


# Choose the matrix from above that you want to plot
mat <- matTop
mat <- matDE1
mat <- matDE2

anno <- as.data.frame(colData(rld)[, c("treat", "stage")])
pheatmap(mat, annotation_col = anno)

#Use heatmap.2 function to draw a heatmap
png(file="TopGene.heatmap.png",width=500,height=500,res=72)
heatmap.2( matTop, scale="row",trace="none", dendrogram="column",col = colorRampPalette( rev(brewer.pal(9, "RdBu")) )(255))
dev.off()

#PLOT PCA
png(file="pca_stock.png",width=1000,height=1000,res=72)
print(plotPCA(rld, intgroup=c("treat")))
dev.off()


# creates a own color palette from white to blue
my_palette <- colorRampPalette(c("darkblue", "blue", "cornsilk", "red", "darkred"))(n = 299)


# creates a 5 x 5 inch image
png("./de_genes_padj10_heatmap.png",    # create PNG for the heat map        
    width = 5*300,        # 5 x 300 pixels
    height = 5*300,
    res = 300,            # 300 pixels per inch
    pointsize = 8)        # smaller font size

heatmap.2(matDE1,
          main = "RNA-seq", # heat map title
          density.info="none",  # turns off density plot inside color legend
          trace="none",         # turns off trace lines inside the heat map
          margins =c(12,9),     # widens margins around plot
          col=my_palette)       # use on color palette defined earlier
dev.off()               # close the PNG device


# creates a 5 x 5 inch image
png("./de_genes_p-unadj05_heatmap.png",    # create PNG for the heat map        
    width = 5*300,        # 5 x 300 pixels
    height = 5*300,
    res = 300,            # 300 pixels per inch
    pointsize = 8)        # smaller font size

heatmap.2(matDE2,
          main = "RNA-seq", # heat map title
          density.info="none",  # turns off density plot inside color legend
          trace="none",         # turns off trace lines inside the heat map
          margins =c(12,9),     # widens margins around plot
          col=my_palette)       # use on color palette defined earlier
dev.off()               # close the PNG device


#VOLCANO PLOTTING
# Choose specific contrast above and produce "res" table for plotting log2fc and pvalue
# Compute significance, with a maximum of ~320 for the p-values set to 0 due to limitation of computation precision
res$sig <- -log10(res$pvalue)
sum(is.infinite(res$sig))
res[is.infinite(res$sig),"sig"] <- 350
# View(res[is.na(res$pvalue),])

# Select genes with a defined p-value (DESeq2 assigns NA to some genes)
genes.to.plot <- !is.na(res$pvalue)
# sum(genes.to.plot)
range(res[genes.to.plot, "log2FoldChange"])

## Volcano plot of adjusted p-values
cols <- densCols(res$log2FoldChange, res$sig)
cols[res$pvalue ==0] <- "purple"
res$pch <- 19
res$pch[res$pvalue ==0] <- 6
plot(res$log2FoldChange, 
     res$sig, 
     col=cols, panel.first=grid(),
     main="Volcano plot", 
     xlab="Effect size: log2(fold-change)",
     ylab="-log10(adjusted p-value)",
     pch=res$pch, cex=0.4)
abline(v=0)
abline(v=c(-2,2), col="brown")
abline(h=-log10(alpha), col="brown")

## Plot the names of a reasonable number of genes, by selecting those begin not only significant but also having a strong effect size
gn.selected <- abs(res$log2FoldChange) > 1.500 & res$pvalue < alpha 
text(res$log2FoldChange[gn.selected],
     -log10(res$pvalue)[gn.selected],
     lab=rownames(res)[gn.selected ], cex=0.6)
